js/store.js

```diff
@@ -134,19 +256,74 @@ const store = {
     const firstParty = this.getWebsite(origin);
     const thirdParty = this.getWebsite(target);
 
+/*
+  if not {},
+
+  firstParty is expected to match this format:
+  {
+    thirdPartyHostnames: [
+      "www.thirdpartydomain.com",
+      ...
+    ]
+  }
+  thirdParty is expected to match this format:
+  {
+    firstPartyHostnames: [
+      "www.firstpartydomain.com",
+      ...
+    ],
+    isIgnored: true, (optional)
+    whiteListedFirstParty: "www.google.com" (optional)
+  }
+  */
+
+    let whiteListOverride = false;
+    let onWhiteList = false;
+
     if (!('thirdPartyHostnames' in firstParty)) {
       firstParty['thirdPartyHostnames'] = [];
     }
     if (!firstParty['thirdPartyHostnames'].includes(target)) {
-      firstParty['thirdPartyHostnames'].push(target);
-      newThirdParty = true;
+      // third party is not currently linked to first party
+      if (!thirdParty['isIgnored']) {
+        // third party needs a whitelist check
+        onWhiteList = this.onWhiteList(origin, target);
+        if (onWhiteList) {
+          // third party is whitelisted for this first party
+          // ignore the link for now
+          thirdParty['isIgnored'] = true;
+          thirdParty['whiteListedFirstParty'] = origin;
+        } else {
+          // third party is not whitelisted for this first party
+          firstParty['thirdPartyHostnames'].push(target);
+          newThirdParty = true;
+        }
+      } else if (!(this.onWhiteList(origin, target))) {
+        // third party was previously whitelisted, and its
+        // previously ignored link with the owning first party
+        // should be added now that it links to a first party
+        // for which it is not whitelisted
+        thirdParty['isIgnored'] = false;
+        whiteListOverride = true;
+        const whiteListedFirstParty = thirdParty['whiteListedFirstParty'];
+        this._websites[whiteListedFirstParty]['thirdPartyHostnames']
+          .push(target);
+        firstParty['thirdPartyHostnames'].push(target);
+        // take care of updating thirdParty[firstPartyHostnames] further down
+        newThirdParty = true;
+      }
     }
-    if (!('firstPartyHostnames' in firstParty)) {
+    if (!('firstPartyHostnames' in thirdParty)) {
       thirdParty['firstPartyHostnames'] = [];
     }
     if (!thirdParty['firstPartyHostnames'].includes(origin)) {
-      thirdParty['firstPartyHostnames'].push(origin);
-      newThirdParty = true;
+      if (whiteListOverride) {
+        thirdParty['firstPartyHostnames']
+          .push(thirdParty['whiteListedFirstParty']);
```

Comment:
`thirdParty['whiteListedFirstParty']` is only the same as `origin` in the initial third party request for which the third party is whitelisted by the request's first party.

In this if statement, `origin` is a new first party for which the third party is not whitelisted, so `origin` is not equal to `thirdParty['whiteListedFirstParty']`.